horovod/run/run_func.py

```diff
@@ -0,0 +1,199 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+from __future__ import print_function
+import cloudpickle
+import tempfile
+import os
+import sys
+import six
+import subprocess
+import textwrap
+
+from horovod.run.common.util import safe_shell_exec
+
+_PICKLED_RESULT_PREFIX = "result.pkl"
+
+_LOCAL_PICKLED_PROC_FN_FILENAME = "local_proc_fn.pkl"
+_PICKLED_PROC_FN_FILENAME = "proc_fn.pkl"
+
+_LOCAL_PROC_LAUNCHER_FILENAME = "local_launch.sh"
+_PROC_LAUNCHER_FILENAME = "launch.sh"
+
+
+def run_func(
+        fn,
+        num_proc,
+        host=None,
+        ssh_port=None,
+        disable_cache=False,
+        start_timeout=None,
+        env=None,
+        verbose=False):
+    """
+    Run horovod inside python program.
+    Run `num_proc` processes executing `fn` on specified hosts.
+
+    :param fn: Function to run.
+    :param num_proc: Number of Horovod processes.
+    :param host: To specify the list of host names as well as the
+                 number of available slots on each host for
+                 training processes using the following format:
+                 <hostname>:<number of slots>,... .
+                 E.g., host1:2,host2:4,host3:1 indicates that 2 processes can run on
+                 host1, 4 processes on host2, and 1 process on host3.
+                 If None, launch all mpi processes on local node.
+    :param ssh_port: SSH port on all the hosts.
+                     If None, use default SSH port.
+    :param disable_cache: If the flag is not set, horovodrun will perform
+                          the initialization checks only once every 60
+                          minutes -- if the checks successfully pass.
+                          Otherwise, all the checks will run every time
+                          horovodrun is called."
+    :param start_timeout: Horovodrun has to perform all the checks and
+                          start the processes before the specified
+                          timeout. The default value is 30 seconds.
+                          Alternatively, The environment variable
+                          HOROVOD_START_TIMEOUT can also be used to
+                          specify the initialization timeout.
+    :param env: Environment dictionary to use in Horovod run.
+                If None, use `os.environ`.
+    :param verbose: If this flag is set, extra messages will be printed.
+
+    :return: A list contains returned result from each MPI process. The ith value of the list
+             corresponds to the returned result of rank i process.
+    """
+
+    wdir = tempfile.mkdtemp()
+    if verbose:
+        print("horovod.run_func working dir is " + wdir)
+
+    if ssh_port:
+        ssh_port_opt = "-p " + str(ssh_port)
+        scp_port_opt = "-P " + str(ssh_port)
```

Comment:
no need to define `scp_port_opt`, it is the same as SSH port

Reply:
oh, they're different. scp should only use "-P", ssh on some system only support "-p", some system support both "-p" and "-P". So for safety ssh I use "-p"